### PR TITLE
packages: support custom queries

### PR DESCRIPTION
**Pull request by uint**
In order to resolve https://github.com/confio/poe-contracts/issues/36, it looks like we need the `tgrade-valset` contract to use `TgradeQuery` as the custom query type for `Deps` and `DepsMut` (currently it's `Empty`).

In order to do that, `Admin` needs to support custom query types that are not `Empty` in `Deps`/`DepsMut`.

### CODE REVIEW

**Comment by ethanfrey** [comment]
Ah, yes, now that I made Deps/DepsMut generic over the custom query type, we need to make controllers generic over that as well, not just Response<T>.

I agree. Can you also update other controllers? (Possibly as a follow up)

**Review by ethanfrey — APPROVED** [review]
nice find

**Comment by ethanfrey** [comment]
Note that main is breaking, so after merge, you should make a new branch off of v0.11.1, back port this fix (and for the other controllers) and release a v0.11.2.

I don't think v0.12.0 is coming out so fast, so let's just patch the last stable version

**Comment by uint** [comment]
> I agree. Can you also update other controllers? (Possibly as a follow up)

Happy to do that, should be easy enough.

**Comment by maurolacy** [comment]
OK with me. Only thing is that v0.12 will include type safe bounds, and it will require code migration.

Migration should be easy, though. And, there's already a MIGRATING entry for this.

**Review by maurolacy — APPROVED** [review]
LGTM.

**Comment by maurolacy** [comment]
As @ethanfrey says, you can either cherry pick this into a `0.11` branch (branched off from the latest 0.11 version tag) and release a new 0.11, or release 0.12 early. I for one would do the 2nd, even when it'll require some code migration on the client repo.

**Comment by ethanfrey** [comment]
Speaking of.. let me tag the v0.12.0-alpha2 release I prepared.
And then I will make one more PR I would like in a final v0.12.0. I will try to finish tonight, be able to review/merge tomorrow

**Comment by maurolacy** [comment]
Great. I can help with releasing 0.12 final if needed.

**Comment by uint** [comment]
I realized we might also need to update multitest with custom queries and I'm taking a shot at that, but it's not trivial.

**Comment by ethanfrey** [comment]
This is the relevant PR: https://github.com/CosmWasm/cw-plus/pull/653

I would like that in before a v0.12.0 is tagged

**Comment by uint** [comment]
Okay, I tried to cover all of the `/packages`, `multi-test` included. Hopefully I didn't miss anything. Can you please take another look?

**Comment by uint** [comment]
> As @ethanfrey says, you can either cherry pick this into a 0.11 branch (branched off from the latest 0.11 version tag) and release a new 0.11, or release 0.12 early. I for one would go for the 2nd, even when it'll require some code migration on the client repo(s).

I'd rather cut the new release than backport things. I guess for the migration there's nothing to do on my side if there's already a `MIGRATING.md` entry? Just that this is breaking for users?

File: packages/multi-test/src/lib.rs
Comment (maurolacy) [inline packages/multi-test/src/lib.rs:9]: I understand this is needed on all three modules below, right?

File: packages/controllers/src/admin.rs
Comment (maurolacy) [inline packages/controllers/src/admin.rs:37]: Why not defining `Q` over the struct? That way it can have a default of `Empty`.

You would require a `PhantomData`, but it's probably worth it.

The same with the other structs.

Comment (maurolacy) [reply packages/controllers/src/admin.rs:37]: It seems doing that is more work than I anticipated... your call.

File: packages/multi-test/src/contracts.rs
Comment (maurolacy) [inline packages/multi-test/src/contracts.rs:14]: 👍🏼

File: packages/multi-test/src/contracts.rs
Comment (maurolacy) [inline packages/multi-test/src/contracts.rs:140]: Type complexity anyone?

File: packages/multi-test/src/lib.rs
Comment (uint) [reply packages/multi-test/src/lib.rs:9]: No, just `contracts`. We can make it more granular if we want to. The lint seems like it's sometimes broken though:

https://github.com/rust-lang/rust-clippy/issues/1013#issuecomment-587054810

File: packages/controllers/src/admin.rs
Comment (uint) [reply packages/controllers/src/admin.rs:37]: I'll take a look!

**Review by maurolacy — APPROVED** [review]
LGTM.

File: packages/multi-test/src/lib.rs
Comment (maurolacy) [reply packages/multi-test/src/lib.rs:9]: Let's do that if possible, yes. That will give us the opportunity to at least flag upcoming and take a look at future type complexities.

File: packages/controllers/src/admin.rs
Comment (uint) [reply packages/controllers/src/admin.rs:37]: I wonder about this. From another point of view, I feel like `Admin` shouldn't be concerned with what the custom query type is.

And I'm not sure the default of `Empty` is such a big benefit - in typical calls of these functions type inference should do the trick without the need to write something like `set::<TgradeQuery>(...)`... I think?

Comment (uint) [reply packages/controllers/src/admin.rs:37]: For example, this line still works just fine after adding the type parameter to `set`:

https://github.com/CosmWasm/cw-plus/blob/controllers-custom-query-type/contracts/cw20-ics20/src/contract.rs#L42

Comment (uint) [reply packages/controllers/src/admin.rs:37]: Yeah, I'm keeping this as is.

File: packages/multi-test/src/contracts.rs
Comment (uint) [reply packages/multi-test/src/contracts.rs:140]: Are you suggesting simplifying this somehow? I'd definitely like to, I just have no idea how as of right now.

File: packages/multi-test/src/lib.rs
Comment (uint) [reply packages/multi-test/src/lib.rs]: That's fair, done.

File: packages/multi-test/src/contracts.rs
Comment (maurolacy) [reply packages/multi-test/src/contracts.rs:140]: No. Just couldn't help noticing the sheer number of type params.

Comment (uint) [reply packages/multi-test/src/contracts.rs:140]: Ah, yeah, it's... something :grin:

**Pull request merged**
